### PR TITLE
#160452492 Implement article sharing by email

### PR DESCRIPTION
**Pull request by olusoladavid**
#### What does this PR do?
- Implement article sharing by email

### Description of Task to be completed?
- Write tests
- Create mailto endpoint
- Add private/public mode to article model
- Add published/unpublished mode to article model
- Send mail to share action recipient

#### How should this be manually tested?
- Fetch and checkout to the branch
```$xslt
> git fetch
> git checkout ft-160452492-share-articles-by-email
> npm install 
```
- Setup database using .env-sample
* Run Sequelize migration 
```$xslt
 > npm run migration
 ```
- Run application
```$xslt 
> npm run start
```
- To run test
```$xslt 
> npm run test
```

* Using a REST client such as postman to test endpoints: 
To share an article by email: `POST /api/v1/articles/{slug}/mailto` 

sample request body
```JSON
{
    "email": "arealemail@gmail.com"
}
```
* For API Documentation
http://localhost:3000/api-docs/

#### What are the relevant pivotal tracker stories?
- [#160452492](https://www.pivotaltracker.com/story/show/160452492)

### CODE REVIEW

File: controllers/article/ShareController.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+   * @param {Object} req Request object
+   * @param {Object} res Response object
+   * @param {Function} next Next middleware
+   * @returns {null} Calls the next middleware
+   * @memberof ShareController
+   */
+  static async validateInputs(req, res, next) {
+    const { email: recipient } = req.body;
+    if (!recipient || !validator.isEmail(recipient)) {
+      return res.status(400).json({
```
Comment (blackshady) [inline controllers/article/ShareController.js]: Nice job on your Implementation of this feature. I like the way you try to follow SRP. I would suggest you use the ```throwErrorIfNull()``` method and pass in the HTTP status code. What do you think?

Comment (olusoladavid) [reply controllers/article/ShareController.js]: Thank you for the catch @blackshady . I will refactor the code to reuse the `throwErrorIfNull()` utility.

File: services/Mail.js [hunk 2]
```diff
@@ -139,16 +137,39 @@ class Mail {
...
-     */
+   * Method to complete article sharing by email
+   * @static
+   * @param {object} articleShareData
+   * @return {json} Returns json object
+   * @memberof Mail
+   */
+  static async shareArticle(articleShareData) {
+    const {
+      article: { slug, title },
```
Comment (blackshady) [inline services/Mail.js:148]: Nice job using deep object destructing.

File: tests/routes/emailShare.test.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import { defaultArticle } from '../testHelpers/articleUtil';
+import { deleteTable } from '../testHelpers';
+
+const { expect } = chai;
+const { User, Article } = db;
+
+describe('POST /share', () => {
+  let user, jwt, publicArticleSlug, privateArticleSlug, unpublishedArticleSlug;
+
+  before(async () => {
```
Comment (blackshady) [inline tests/routes/emailShare.test.js:17]: It is sometimes advisable to add some sort of message to your test hook, this is to make it easy to debug once there is an error.

File: tests/routes/emailShare.test.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  });
+
+  it("should send a mail to an authenticated user's contact", async () => {
+    const response = await chai
+      .request(server)
+      .post(`/api/v1/articles/${publicArticleSlug}/mailto`)
+      .set(AUTHORIZATION_HEADER, `Bearer: ${jwt}`)
+      .send({ email: 'olusola.oguntimehin@andela.com' });
+    expect(response).to.have.status(200);
+    // expect(mailServer).to.be.called();
```
Comment (blackshady) [inline tests/routes/emailShare.test.js]: Please remove this line of code if it is not needed

File: services/Mail.js (hunk 2 above)
Comment (olusoladavid) [reply services/Mail.js:148]: Thank you.

File: tests/routes/emailShare.test.js (hunk 3 above)
Comment (olusoladavid) [reply tests/routes/emailShare.test.js:17]: Thank you. Mocha handles this out of the box by specifying that the error occurred in the `before` hook.

File: tests/routes/emailShare.test.js (hunk 4 above)
Comment (olusoladavid) [reply tests/routes/emailShare.test.js]: Thank you for the catch. I will correct the oversight.

**Review by Alameen688 — APPROVED** [review]
Looks good to me. Works as expected.

**Pull request merged**